core/runtime/src/main/scala/com/fortysevendeg/exercises/Exercises.scala

```diff
@@ -0,0 +1,64 @@
+package com.fortysevendeg.exercises
+
+import scala.annotation.tailrec
+import scala.collection.JavaConversions._
+
+import java.net.URL
+import java.nio.charset.Charset
+import java.io.BufferedReader
+import java.io.InputStreamReader
+import java.lang.ClassNotFoundException
+import java.nio.charset.StandardCharsets
+
+import cats._
+import cats.data.Xor
+
+object Exercises {
+
+  private[exercises] val LIBRARY_FILE_PATH = "scala-exercises/library.47"
+
+  def discoverLibraries(cl: ClassLoader = classOf[Exercise].getClassLoader) = {
+
+    /** Reads the lines, in reverse order, from a URL resource. */
+    def readLinesReverse(url: URL): List[String] = {
+      val in = new BufferedReader(new InputStreamReader(url.openStream, StandardCharsets.UTF_8))
+      @tailrec def accLines(lines: List[String] = Nil): List[String] = {
+        val line = in.readLine()
+        if (line eq null) lines else accLines(line :: lines)
+      }
+      accLines()
+    }
+
+    val classNames = cl.getResources(LIBRARY_FILE_PATH)
+      .foldLeft(Nil: List[String]) { (acc, url) ⇒ readLinesReverse(url) ++ acc }
+      .reverse
+      .map(_.trim)
+      .filterNot(_.isEmpty)
+
+    type Acc = (List[String], List[Library])
```

Comment:
That's fine nothing wrong with your version. If we end up needing to use those return types in monadic comprehensions or with applicative builders for whatever reasons then we can port it to one of those types.